clippy_utils/src/res.rs

```diff
@@ -0,0 +1,643 @@
+use rustc_span::{Ident, Symbol};
+
+/// Either a `HirId` or a type which can be identified by one.
+pub trait HasHirId: Copy {
+    fn hir_id(self) -> HirId;
+}
+impl HasHirId for HirId {
+    #[inline]
+    fn hir_id(self) -> HirId {
+        self
+    }
+}
+impl HasHirId for &Expr<'_> {
+    #[inline]
+    fn hir_id(self) -> HirId {
+        self.hir_id
+    }
+}
+
+type DefRes = (DefKind, DefId);
+
+pub trait MaybeTypeckRes<'tcx> {
+    /// Gets the contained `TypeckResults`.
+    ///
+    /// With debug assertions enabled this will always return `Some`. `None` is
+    /// only returned so logic errors can be handled by not emitting a lint on
+    /// release builds.
+    fn typeck_res(&self) -> Option<&TypeckResults<'tcx>>;
+
+    /// Gets the type-dependent resolution of the specified node.
+    ///
+    /// With debug assertions enabled this will always return `Some`. `None` is
+    /// only returned so logic errors can be handled by not emitting a lint on
+    /// release builds.
+    #[inline]
+    #[cfg_attr(debug_assertions, track_caller)]
+    fn ty_based_def(&self, node: impl HasHirId) -> Option<DefRes> {
+        #[inline]
+        #[cfg_attr(debug_assertions, track_caller)]
+        fn f(typeck: &TypeckResults<'_>, id: HirId) -> Option<DefRes> {
+            if typeck.hir_owner == id.owner {
+                let def = typeck.type_dependent_def(id);
+                debug_assert!(
+                    def.is_some(),
+                    "attempted type-dependent lookup for a node with no definition\
+                        \n  node `{id:?}`",
+                );
+                def
+            } else {
+                debug_assert!(
+                    false,
+                    "attempted type-dependent lookup for a node in the wrong body\
+                        \n  in body `{:?}`\
+                        \n  expected body `{:?}`",
+                    typeck.hir_owner, id.owner,
+                );
+                None
+            }
+        }
+        self.typeck_res().and_then(|typeck| f(typeck, node.hir_id()))
+    }
+}
+impl<'tcx> MaybeTypeckRes<'tcx> for LateContext<'tcx> {
+    #[inline]
+    #[cfg_attr(debug_assertions, track_caller)]
+    fn typeck_res(&self) -> Option<&TypeckResults<'tcx>> {
+        if let Some(typeck) = self.maybe_typeck_results() {
+            Some(typeck)
+        } else {
+            // It's possible to get the `TypeckResults` for any other body, but
+            // attempting to lookup the type of something across bodies like this
+            // is a good indication of a bug.
+            debug_assert!(false, "attempted type-dependent lookup in a non-body context");
+            None
+        }
+    }
+}
+impl<'tcx> MaybeTypeckRes<'tcx> for TypeckResults<'tcx> {
+    #[inline]
+    fn typeck_res(&self) -> Option<&TypeckResults<'tcx>> {
+        Some(self)
+    }
+}
+
+/// A `QPath` with the `HirId` of the node containing it.
+type QPathId<'tcx> = (&'tcx QPath<'tcx>, HirId);
```

Comment:
Could be made `pub` since it is used in pub trait method signatures.

Reply:
These aliases only exist to make the trait signatures shorter since they're written quite a few times. Outside the module it shouldn't even come up and if it does it will be written once.